server/controllers/stock/import.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * Stock Import Module
+ *
+ * This module is responsible of handling the import of stock
+ * and related stock quantities
+ */
+const path = require('path');
+const uuid = require('uuid');
+const moment = require('moment');
+
+const db = require('../../lib/db');
+const util = require('../../lib/util');
+const BadRequest = require('../../lib/errors/BadRequest');
+
+exports.downloadTemplate = downloadTemplate;
+exports.importStock = importStock;
+
+/**
+ * @method downloadTemplate
+ *
+ * @description send to the client the template file for stock import
+*/
+function downloadTemplate(req, res, next) {
+  try {
+    const file = path.join(__dirname, '../../resources/templates/import-stock-template.csv');
+    res.download(file);
+  } catch (error) {
+    next(error);
+  }
+}
+
+/**
+ * @method importStock
+ *
+ * @description this method allow to do an import of stock and their lots
+ */
+function importStock(req, res, next) {
+  if (!req.files || req.files.length === 0) {
+    const errorDescription = 'Expected at least one file upload but did not receive any files.';
+    const errorDetails = new BadRequest(errorDescription, 'ERRORS.MISSING_UPLOAD_FILES');
+    next(errorDetails);
+    return;
+  }
+
+  let query;
+  let queryParams;
+
+  const filePath = req.files[0].path;
+  const depotUuid = db.bid(req.body.depot_uuid);
+  const documentUuid = db.bid(uuid.v4());
+
+  // be sure that the depot exist
+  db.one('SELECT uuid FROM depot WHERE uuid = ?', depotUuid)
+    .then(() => {
+      return util.formatCsvToJson(filePath);
+    })
+    .then(data => {
+      if (!hasValidDataFormat(data)) {
+        throw new BadRequest('The given file has a bad data format for stock', 'ERRORS.BAD_DATA_FORMAT');
+      }
+
+      const transaction = db.transaction();
+
+      data.forEach(item => {
+        query = 'CALL ImportStock(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);';
+        queryParams = [
+          req.session.enterprise.id,
+          req.session.project.id,
+          req.session.user.id,
+          depotUuid,
+          documentUuid,
+          item.inventory_group_name,
+          item.inventory_code || '',
+          item.inventory_text,
+          item.inventory_type,
+          item.inventory_unit,
+          item.inventory_unit_price,
+          item.inventory_cmm || 0,
+          item.stock_lot_label,
+          item.stock_lot_quantity,
+          moment(item.stock_lot_expiration).format('YYYY-MM-DD'),
+        ];
+        transaction.addQuery(query, queryParams);
```

Comment:
I just thought of this - but most of our other modules that write multiple lines in the database use `Staging` tables ... such as [invoicing](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/procedures/invoicing.sql#L15), cash, and vouchers.  In theory, this gives us both a performance improvement (since we can import all the data at one time) and better error handling (since we can say that inventory XYZ doesn't exist before we even start to import the data).

It would be worth writing a second version of this function that uses temporary tables to compare with this version.  That doesn't need to be done for this PR, but we should do some performance analysis just to see.